ast/src/parsed/mod.rs

```diff
@@ -633,6 +635,51 @@ pub enum BinaryOperator {
     Greater,
 }
 
+#[derive(Debug, PartialEq, Eq)]
+pub enum BinaryOperatorAssociativity {
+    Left,
+    Right,
+}
+
+impl BinaryOperator {
+    pub fn precedence(&self) -> ExpressionPrecedence {
+        use BinaryOperator::*;
+        match self {
+            // Unary - * ! & &mut
+            Pow => 2,
+            // * / %
+            Mul | Div | Mod => 3,
+            // + -
+            Add | Sub => 4,
+            // << >>
+            ShiftLeft | ShiftRight => 5,
+            // &
+            BinaryAnd => 6,
+            // ^
+            BinaryXor => 7,
+            // |
+            BinaryOr => 8,
+            // == != < > <= >=
+            Equal | NotEqual | Less | Greater | LessEqual | GreaterEqual => 9,
+            // &&
+            LogicalAnd => 10,
+            // ||
+            LogicalOr => 11,
+            // .. ..=
+            // ??
+            // = += -= *= /= %= &= |= ^= <<= >>=
+            Identity => 12,
+        }
+    }
+
+    pub fn associativity(&self) -> BinaryOperatorAssociativity {
+        match self {
+            BinaryOperator::Pow => BinaryOperatorAssociativity::Right,
```

Comment:
I'm not sure if this is correct. I'm also not sure if your grammar is correct. Can you check that `x**y**2` parses the same way as `x**(y**2)`?

Reply:
If not we have to change the grammar.

Reply:
We forgot to include identity (= += -= *= /= %= &= |= ^= <<= >>=) as right associative.
Also there is an imporatnt thing, the associativey of some fiels is "Require parentheses" ! we must handle it as well
https://doc.rust-lang.org/reference/expressions.html#expression-precedence

Reply:
I found this in pilcom source code, I assume power is left assosiative and we must do the same
https://github.com/0xPolygonHermez/pilcom/blob/36bfb6ab4e0f1bae3bbeda11cc6305ede153f6b6/src/pil_parser.jison#L51
@chriseth can you double check?

Reply:
It's a good thing that the parser cannot parse `x = y <= z`. The reason is because `=` and `<=` have the same precedence. Please add tests for `(x = y) <= z` and `x = (y <= z)`.